applications/GeoMechanicsApplication/custom_elements/Pw_element.h

```diff
@@ -0,0 +1,601 @@
+
+        std::vector<Vector> projected_gravity;
+        projected_gravity.reserve(number_integration_points);
+        if (GetGeometry().LocalSpaceDimension() == 1) {
+            for (unsigned int integration_point_index = 0;
+                 integration_point_index < number_integration_points; ++integration_point_index) {
+                GeoElementUtilities::InterpolateVariableWithComponents<TDim, TNumNodes>(
+                    body_acceleration, rNContainer, volume_acceleration, integration_point_index);
+                array_1d<double, TDim> tangent_vector = column(J_container[integration_point_index], 0);
+                tangent_vector /= norm_2(tangent_vector);
+                projected_gravity.emplace_back(
+                    ScalarVector(1, std::inner_product(tangent_vector.begin(), tangent_vector.end(),
+                                                       body_acceleration.begin(), 0.0)));
+            }
+        } else {
+            for (unsigned int integration_point_index = 0;
+                 integration_point_index < number_integration_points; ++integration_point_index) {
+                GeoElementUtilities::InterpolateVariableWithComponents<TDim, TNumNodes>(
+                    body_acceleration, rNContainer, volume_acceleration, integration_point_index);
+                projected_gravity.emplace_back(body_acceleration);
+            }
+        }
+
+        return projected_gravity;
+    }
+
+    std::unique_ptr<IntegrationCoefficientModifier> CloneIntegrationCoefficientModifier() const
+    {
+        return mIntegrationCoefficientsCalculator.CloneModifier();
+    }
+
+    std::unique_ptr<ContributionCalculator<TNumNodes>> CreateCalculator(const CalculationContribution& rContribution,
+                                                                        const ProcessInfo& rCurrentProcessInfo)
+    {
+        switch (rContribution) {
+            using enum CalculationContribution;
+        case Permeability:
+            return std::make_unique<PermeabilityCalculator<TNumNodes>>(CreatePermeabilityInputProvider());
+        case Compressibility:
+            if (GetProperties()[RETENTION_LAW] == "PressureFilterLaw") {
+                return std::make_unique<FilterCompressibilityCalculator<TNumNodes>>(
+                    CreateFilterCompressibilityInputProvider(rCurrentProcessInfo));
+            }
+            return std::make_unique<CompressibilityCalculator<TNumNodes>>(
+                CreateCompressibilityInputProvider(rCurrentProcessInfo));
+        case FluidBodyFlow:
+            return std::make_unique<FluidBodyFlowCalculator<TNumNodes>>(CreateFluidBodyFlowInputProvider());
+        default:
+            KRATOS_ERROR << "Unknown contribution" << std::endl;
+        }
+    }
+
+    void CachingDataForCalculator()
+    {
+        mIntegrationCoefficients = SaveIntegrationCoefficients();
+        mNContainer              = SaveNContainer();
+        mFluidPressures          = SaveFluidPressure();
```

Comment:
Would it be possible to use CalculateIntegrationCoefficients here? Since the functions don't really save the variables, but just calculate them, I'd propose to reflect that in the name:

```suggestion
        mIntegrationCoefficients = CalculateIntegrationCoefficients();
        mNContainer              = CalculateNContainer();
        mFluidPressures          = CalculateFluidPressures();
```

Reply:
changed names as you suggested